libc/src/math/generic/range_reduction_double_common.h

```diff
@@ -0,0 +1,173 @@
+  // For FMA:
+  //   | {x * 128/pi} - (y_hi + y_lo) | <= 2 * 2^77 * 2^-103 * 2^-52
+  //                                    =  2^-77.
+  //   | {x mod pi/128} - (u.hi + u.lo) | < 2 * 2^-6 * 2^-77.
+  //                                      = 2^-82.
+  // For non-FMA:
+  //   | {x * 128/pi} - (y_hi + y_lo) | <= 2 * 2^77 * 2^-99 * 2^-52
+  //                                    =  2^-73.
+  //   | {x mod pi/128} - (u.hi + u.lo) | < 2 * 2^-6 * 2^-73.
+  //                                      = 2^-78.
+  u = fputil::quick_mult<NO_FMA>(y, PI_OVER_128_DD);
+
+  return static_cast<unsigned>(static_cast<int64_t>(kh) +
+                               static_cast<int64_t>(km));
+}
+
+LIBC_INLINE Float128 range_reduction_small_f128(double x) {
+  double prod_hi = x * ONE_TWENTY_EIGHT_OVER_PI[3][0];
+  double kd = fputil::nearest_integer(prod_hi);
+
+  Float128 mk_f128(-kd);
+  Float128 x_f128(x);
+  Float128 p_hi =
+      fputil::quick_mul(x_f128, Float128(ONE_TWENTY_EIGHT_OVER_PI[3][0]));
+  Float128 p_mid =
+      fputil::quick_mul(x_f128, Float128(ONE_TWENTY_EIGHT_OVER_PI[3][1]));
+  Float128 p_lo =
+      fputil::quick_mul(x_f128, Float128(ONE_TWENTY_EIGHT_OVER_PI[3][2]));
+  Float128 s_hi = fputil::quick_add(p_hi, mk_f128);
+  Float128 s_lo = fputil::quick_add(p_mid, p_lo);
+  Float128 y = fputil::quick_add(s_hi, s_lo);
+
+  return fputil::quick_mul(y, PI_OVER_128_F128);
+}
+
+// TODO: Maybe not redo-ing most of the computation, instead getting
+//   y_hi, idx, pm.lo, x_reduced from range_reduction_large.
+template <bool NO_FMA>
+LIBC_INLINE Float128 range_reduction_large_f128(double x) {
+  // |x| >= 2^32 for FMA and |x| >= 2^23 for non-FMA.
+  using FPBits = typename fputil::FPBits<double>;
+  FPBits xbits(x);
+
+  int x_e_m62 = xbits.get_biased_exponent() - (FPBits::EXP_BIAS + 62);
+  int idx = (x_e_m62 >> 4) + 3;
+  // Scale x down by 2^(-(16 * (idx - 3))
+  xbits.set_biased_exponent((x_e_m62 & 15) + FPBits::EXP_BIAS + 62);
+  double x_reduced = xbits.get_val();
+  // x * c_hi = ph.hi + ph.lo exactly.
+  DoubleDouble ph =
+      fputil::exact_mult<NO_FMA>(x_reduced, ONE_TWENTY_EIGHT_OVER_PI[idx][0]);
+  // x * c_mid = pm.hi + pm.lo exactly.
+  DoubleDouble pm =
+      fputil::exact_mult<NO_FMA>(x_reduced, ONE_TWENTY_EIGHT_OVER_PI[idx][1]);
+  // Extract integral parts and fractional parts of (ph.lo + pm.hi).
+  double kh = fputil::nearest_integer(ph.lo);
+  double ph_lo_frac = ph.lo - kh; // Exact
+  double km = fputil::nearest_integer(pm.hi + ph_lo_frac);
+  double pm_hi_frac = pm.hi - km; // Exact
+  // x * 128/pi mod 1 ~ y_hi + y_lo = u.hi + u.lo
+  double y_hi = ph_lo_frac + pm_hi_frac; // Exact
+  // y_lo = x * c_lo + pm.lo
+  Float128 y_lo_0(x_reduced * ONE_TWENTY_EIGHT_OVER_PI[idx][3]);
+  Float128 y_lo_1 = fputil::quick_mul(
+      Float128(x_reduced), Float128(ONE_TWENTY_EIGHT_OVER_PI[idx][2]));
+  Float128 y_lo_2(pm.lo);
+  Float128 y_hi_f128(y_hi);
+
+  using fputil::quick_add;
```

Comment:
This `using` statement doesn't help us minimize the line length on L164 to avoid wrapping. I'd drop it and just qualify `quick_add`.

Reply:
Done.